File: Python/copy_list_random_ptr.py
"""
# Definition for a Node.
class Node:
    def __init__(self, x: int, next: 'Node' = None, random: 'Node' = None):
        self.val = int(x)
        self.next = next
        self.random = random
"""

class Solution:
    def copyRandomList(self, head: 'Optional[Node]') -> 'Optional[Node]':
        if not head:
            return None
        
        record = {}
        node = head
        while node:
            new_node = Node(node.val)
            record[node] = new_node
            node = node.next
        
        node = head
        while node:
            new_node = record[node]
            new_node.next = record[node.next] if node.next else None
            new_node.random = record[node.random] if node.random else None

            node = node.next
        
        return record[head]

File: Cpp/copy_list_random_ptr.cpp
/*
// Definition for a Node.
class Node {
public:
    int val;
    Node* next;
    Node* random;
    
    Node(int _val) {
        val = _val;
        next = NULL;
        random = NULL;
    }
};
*/

class Solution {
public:
    Node* copyRandomList(Node* head) {
        if(head==NULL) return NULL;
        unordered_map<Node*, Node*> nodes;
        Node* start_node = head;

        while(start_node != NULL){
            nodes[start_node] = new Node(start_node -> val);
            start_node = start_node ->next;
        }
        
        Node* st_node = head;
        while(st_node != NULL){
            Node* new_node = nodes[st_node];
            if(st_node -> next != NULL) new_node -> next = nodes[st_node->next];
            else new_node -> next = NULL;

            if(st_node -> random != NULL) new_node -> random = nodes[st_node->random];
            else new_node -> random = NULL;
            st_node = st_node ->next;
        }
        return nodes[head];
    }
};